packages/jest-haste-map/src/crawlers/watchman.js

```diff
@@ -18,126 +18,127 @@ import H from '../constants';
 
-          pairs.forEach(pair => {
-            const root = normalizePathSep(pair.root);
-            const response = pair.response;
-            if ('warning' in response) {
-              console.warn('watchman warning: ', response.warning);
-            }
+      const query = clocks[root]
+        ? // Use the `since` generator if we have a clock available
+          {expression, fields, since: clocks[root]}
+        : // Otherwise use the `suffix` generator
+          {expression, fields, suffix: extensions};
+
+      const response = await cmd('query', root, query);
+      shouldReset = shouldReset || response.is_fresh_instance;
+      watchmanFileResults.set(root, response);
+    }
+
+    // Reset the file map if watchman was restarted and sends us a list of files.
+    if (shouldReset) {
+      files = Object.create(null);
+    }
+
+    for (const [watchRoot, response] of watchmanFileResults) {
+      const fsRoot = normalizePathSep(watchRoot);
+      if ('warning' in response) {
+        console.warn('watchman warning: ', response.warning);
+      }
+      clocks[fsRoot] = response.clock;
+      for (const fileData of response.files) {
+        const name = fsRoot + path.sep + normalizePathSep(fileData.name);
+        if (!fileData.exists) {
+          delete files[name];
+        } else if (!ignore(name)) {
+          const mtime =
+            typeof fileData.mtime_ms === 'number'
+              ? fileData.mtime_ms
+              : fileData.mtime_ms.toNumber();
+          const isOld = data.files[name] && data.files[name][H.MTIME] === mtime;
+          if (isOld) {
+            files[name] = data.files[name];
+          } else {
+            // See ../constants.js
+            files[name] = ['', mtime, 0, []];
+          }
+        }
+      }
+    }
+  } catch (error) {
+    throw new WatchmanError(error);
+  } finally {
+    client.end();
+  }
 
-            clocks[root] = response.clock;
-            response.files.forEach(fileData => {
-              const name = root + path.sep + normalizePathSep(fileData.name);
-              if (!fileData.exists) {
-                delete files[name];
-              } else if (!ignore(name)) {
-                const mtime =
-                  typeof fileData.mtime_ms === 'number'
-                    ? fileData.mtime_ms
-                    : fileData.mtime_ms.toNumber();
-                const isNew =
-                  !data.files[name] || data.files[name][H.MTIME] !== mtime;
-                if (isNew) {
-                  // See ../constants.js
-                  files[name] = ['', mtime, 0, []];
-                } else {
-                  files[name] = data.files[name];
-                }
-              }
-            });
-          });
-        }),
-      )
-      .then(() => {
-        client.end();
-        data.files = files;
-        resolve(data);
-      })
-      .catch(error => {
-        client.end();
-        reject(WatchmanError(error));
-      });
-  });
+  if (clientError) {
```

Comment:
Hoping this approach works and is OKAY. Otherwise I'd need to wrap the whole thing in a `new Promise()` block :(